### DISCUSSION TITLE
Global HiLo generator, when I should return unused range?

### DISCUSSION THREAD

**Discussion by montyclt** [ravendb/ravendb#13891] [Q&A] [answered]
Hi!, I have a global HiLo generator like this:

```cs
var hilo = new AsyncHiLoIdGenerator("DefaultHiLo", documentStore, options.Database, '\0');
documentStore.Conventions.AsyncDocumentIdGenerator = async (_, entity) => (await hilo.GenerateDocumentIdAsync(entity)[11..];
documentStore.OnSessionDisposing += async (_, _) => await hilo.ReturnUnusedRangeAsync();
```

That is working well for me, but I'm not sure if I should return unused range on session disposing or on document store disposing? Why?

**Comment by ayende** [comment, marked as answer]
You should return it on document store end, yes.
The idea is that the `AsyncHiLoIdGenerator` is used for a long time.

What kind of ids are you trying to generate, by hte way?

**Reply by montyclt** [reply]
I'm trying to use the same HiLo in all documents, and dropping the collection part, so I want to have a documents of multiple collections with ID: 1-A, 2-A, 3-A...

**Reply by ayende** [reply]
The collection part is for humans eyes only, we don't actually need it, but we found that it helps alot when working with the data, FYI

**Reply by montyclt** [reply]
Yes, I know, but for my project, I prefer drop it, so this is why I use a global HiLo generator.

**Reply by mcrio** [reply]
Sidenote: 
Some other approach might be looking at the conventions in order to put the desired types into the same collection: https://ravendb.net/docs/article-page/5.3/csharp/client-api/configuration/identifier-generation/global
